{{ cookiecutter.app_name }}/tests/{{ cookiecutter.module_name }}.py

```diff
@@ -0,0 +1,86 @@
+{%- if cookiecutter.test_framework == 'pytest' %}
+import os
+import sys
+import tempfile
+from pathlib import Path
+
+import pytest
+
+
+if __name__ == "__main__":
+    project_path = Path(__file__).parent.parent
+    os.chdir(project_path)
+
+    ##################################################################
+    # WORKAROUND - On Android, we need to explicitly unpack the test
+    # source code into a location where it can be discovered.
+    ##################################################################
+    if hasattr(sys, "getandroidapilevel"):
+        import tests
+
+        def chaquopy_extract_package(pkg):
+            finder = pkg.__loader__.finder
+            for path in pkg.__path__:
+                chaquopy_extract_dir(finder, finder.zip_path(path))
+
+        def chaquopy_extract_dir(finder, zip_dir):
+            for filename in finder.listdir(zip_dir):
+                zip_path = f"{zip_dir}/{filename}"
+                if finder.isdir(zip_path):
+                    chaquopy_extract_dir(finder, zip_path)
+                else:
+                    finder.extract_if_changed(zip_path)
+        chaquopy_extract_package(tests)
+    ##################################################################
+
+    result = pytest.main(
+        [
```

Comment:
Exactly how the test suite interacts with the main app will vary, so rather than running the tests directly in the `__main__` block, it would be better to put this code in a function.

For example, if the function was called `test`, then for apps structured like the one in https://github.com/beeware/toga/pull/1687, the `__main__` block could look something like this:
```
from {{ cookiecutter.module_name }}.app import main
app = main()
main.add_background_task(Thread(target=test).start)
main.main_loop()
```

I don't think we actually want to include this code in the template just now, because running tests on a background thread would require all the asyncio stuff as well, but at least putting the test entry point in a function would make it easier for the user to control how it's called.

Reply:
Good call - I've made this change.